src/audio/volume/volume.c

```diff
@@ -494,13 +422,16 @@ static inline int volume_set_chan(struct comp_dev *dev, int chan,
 		 * the volume transition always uses the topology defined time.
 		 */
 		if (pga->initial_ramp > 0) {
+			ms = !frames ? VOL_RAMP_UPDATE_US :
+			(VOL_RAMP_UPDATE_US * frames / cd->vol_ramp_frames);
+
 			if (constant_rate_ramp && cd->vol_ramp_range > 0)
 				inc = q_multsr_32x32(cd->vol_ramp_range,
-						     VOL_RAMP_STEP_CONST,
+						     VOL_RAMP_STEP_CONST(ms),
```

Comment:
So the float code likely appears here. Can you check it and convert to all-integer fractional math.

Reply:
I think I don't fully understand. The macro expands to Q_CONVERT_FLOAT, which explicitly casts to float. What do you mean by converting to "all-integer fractional math"?

Reply:
The divide by 1000.0 becomes a run-time float divide if the argument to VOL_RAMP_STEP_CONST(x) is not a constant number calculated by pre-processor. Did you look into disassembly how this got handled?

Fractional arithmetic uses mostly fixed multiplication, and more costly fixed division if integer multiply by inverse is not available. I'm not sure this change emitted C library float code but since there was need to adjust memory I wonder if it happened.

Reply:
Also if things get complicated instead of step size compute (where error does accumulate a bit) the actual gain could be computed from time since volume transition start.

Reply:
@singalsu Yes, unfortunately it looks like float code is generated.

Reply:
It's off-topic but I wonder if it would be possible create compile error if argument for Q_CONVERT_FLOAT() is not a constant value. It's a useful macro but hides this risk.